### PR TITLE
When running a 32-bit rustup on an aarch64 CPU, select a 32-bit toolchain

### LINKED ISSUE #3307: rustup installs incorrect version when 64 bit kernel enabled in 32bit OS.
### Problem

When attempting to install Rust in a 32-bit environment with a 64 kernel the wrong version is installed. This is due to rustup determining 64/32 bit from what is available, not what is being used

### Steps

1. Install fresh 32-bit Raspberry Pi OS image
2. Enable the 64-bit kernel (set `arm_64bit=1` in /boot/config.txt)
3. Restart
4. Follow rust installation instructions
5. Rust installed is 64-bit despite user running in a 32-bit environment (rust wont run)

### Possible Solution(s)

`getconf LONG_BIT` is a better solution than `uname -m` for determining whether the user is operating in 64 bit or 32 bit (which is more in line with user specific installation)

### Notes

Execution of rustc after running rustup:
```shell
robtom@robtom-pi:~ $ rustc --version
error: command failed: 'rustc': No such file or directory (os error 2)
```

### Rustup version

```console
rustup 1.25.2 (17db695f1 2023-02-01)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `(error reading rustc version)`
```

### Installed toolchains

```console
Default host: aarch64-unknown-linux-gnu
rustup home:  /home/control/.rustup

stable-aarch64-unknown-linux-gnu (default)
(error reading rustc version)
```

### PR DESCRIPTION
**Pull request by alex**
this mirrors a similar check that exists in rustup-init.sh

fixes #3307

### CODE REVIEW

**Review by rami3l — APPROVED** [review]
Looks good!

Although I originally expected a runtime detection of the userland's bitness, this approach of using the comptime info does seem to be okay, since being able to run a 32-bit `rustup` does imply that the current system at least has a 32-bit libc.

Thus, this should fix the Raspberry Pi issue if the user is indeed running a 32-bit version of `rustup`.

**Comment by rami3l** [comment]
It's okay to merge this PR first anyway that will indeed fix a common use case...

However, I'm still wondering if something similar will happen on other platforms, and if this approach should be further generalized.

cc @rbtcollins

**Comment by rami3l** [comment]
@alex It suddenly occurred to me that some people might be running Pi OS with a 64-bit userland, with [32-bit compatibility mode enabled](https://unix.stackexchange.com/questions/625576/how-to-run-32-bit-armhf-binaries-on-64-bit-arm64-debian-os-on-raspberry-pi), and even have accidentally downloaded a 32-bit `rustup` binary...

So I guess making the bitness check a runtime one might still be better after all. This way, it will also address your concerns around host/target differences even better. What do you think?

**Comment by alex** [comment]
A runtime check would produce the same results, unless you're suggesting we check `/bin/sh` or some system binary?

**Comment by rami3l** [comment]
@alex I just saw where my misunderstanding was, so sorry for not being clear enough previously! 🙏🙏🙏

I did mean to say `/bin/sh` and I think it should be safe to assume that the machine that runs `rustup` has it.

**Comment by alex** [comment]
I think what I'll do is check `/bin/sh`, but if it doesn't exist, fall back to checking the `rustup` binary.

File: src/dist/dist.rs [hunk 1]
```diff
@@ -346,7 +366,13 @@ impl TargetTriple {
...
                 (b"Linux", b"armv7l") => Some("armv7-unknown-linux-gnueabihf"),
                 (b"Linux", b"armv8l") => Some("armv7-unknown-linux-gnueabihf"),
-                (b"Linux", b"aarch64") => Some(TRIPLE_AARCH64_UNKNOWN_LINUX),
+                (b"Linux", b"aarch64") => {
+                    if is_32bit_userspace() {
+                        Some("armv7-unknown-linux-gnueabihf")
+                    } else {
+                        Some(TRIPLE_AARCH64_UNKNOWN_LINUX)
+                    }
+                }
```
Comment (djc) [inline src/dist/dist.rs]: Style nit: I'd probably write it like this, what do you think?

```suggestion
                (b"Linux", b"aarch64") => Some(match is_32bit_userspace() {
                    true => "armv7-unknown-linux-gnueabihf",
                    false => TRIPLE_AARCH64_UNKNOWN_LINUX,
                }),
```

File: src/dist/dist.rs [hunk 2]
```diff
@@ -219,6 +220,25 @@ impl Deref for TargetTriple {
     }
 }
 
+fn is_32bit_userspace() -> bool {
+    // Check if /bin/sh is a 32-bit binary. If it doesn't exist, fall back to
+    // checking if _we_ are a 32-bit binary.
+    // rustup-init.sh also relies on checking /bin/sh for bitness.
+
+    fn inner() -> io::Result<bool> {
```
Comment (djc) [inline src/dist/dist.rs:229]: Maybe consider adding a note that the inner function is used to simplify error handling?

**Review by djc — APPROVED** [review]

File: src/dist/dist.rs (hunk 1 above)
Comment (alex) [reply src/dist/dist.rs]: I think using `match` on bools is weird :-)

Happy to change if you prefer though.

File: src/dist/dist.rs (hunk 2 above)
Comment (alex) [reply src/dist/dist.rs:229]: Done

File: src/dist/dist.rs (hunk 1 above)
Comment (rami3l) [reply src/dist/dist.rs]: @djc Extracting `Some` is indeed better IMHO, but matching a boolean also looks weird to me... The intent is very clear either way (and given that, I'm okay if the style is not consistent all over the repo, it's a very personal thing after all).

Comment (alex) [reply src/dist/dist.rs]: Did the `Some` part, let me me know if you want the `match` part as well.

Comment (djc) [reply src/dist/dist.rs]: I know it's maybe not the majority opinion of the Rust ecosystem (IMO mostly for being unfamiliar), but I think it's objectively better -- one line less, and with both arms laid out right next to each vertically. Anyway, it's fine either way.

Comment (rami3l) [reply src/dist/dist.rs]: @alex Thanks a lot for bearing with us! We are still quite new here after all, actually we have just gained the permissions today 😂

**Pull request merged**